File: 0210-course-schedule-ii/0210-course-schedule-ii.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> res;
        vector<int> incoming(numCourses, 0);
        vector<vector<int>> adjList(numCourses);
        queue<int> q;

        for (auto pre: prerequisites) {
            adjList[pre[1]].push_back(pre[0]);
            incoming[pre[0]]++;
        }

        for (int i = 0; i < numCourses; i++) {
            if (incoming[i] == 0) q.push(i);
        }

        while (!q.empty()) {
            int node = q.front();
            q.pop();
            res.push_back(node);
            for (auto nei: adjList[node]) {
                incoming[nei]--;
                if (incoming[nei] == 0) q.push(nei);
            }
        }
        if (res.size() == numCourses) return res;
        return {};
    }
};

File: 0210-course-schedule-ii/0210-course-schedule-ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        

        adjList = defaultdict(list)

        for crs, pre in prerequisites:
            adjList[crs].append(pre)
        
        order = []
        visited = set()
        visiting = set()

        def dfs(node):
            visiting.add(node)
            for nei in adjList[node]:
                if nei in visited:
                    continue
                if nei in visiting:
                    return False
                if not dfs(nei):
                    return False
            visiting.remove(node)
            visited.add(node)
            order.append(node)
            return True

        for i in range(numCourses):
            if i not in visited:
                if not dfs(i):
                    return []

        return order